src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -2390,6 +2390,126 @@ BOOST_AUTO_TEST_CASE(GradientSequentialLayerTest)
   BOOST_REQUIRE_LE(CheckGradient(function), 1e-4);
 }
 
+/**
+ * WeightNorm layer numerical gradient test.
+ */
+BOOST_AUTO_TEST_CASE(GradientWeightNormLayerTest)
+{
+  // Linear function gradient instantiation.
+  struct GradientFunction
+  {
+    GradientFunction()
+    {
+      input = arma::randu(10, 1);
+      target = arma::mat("1");
+
+      model = new FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>();
+      model->Predictors() = input;
+      model->Responses() = target;
+      model->Add<IdentityLayer<> >();
+
+      Linear<>* linear = new Linear<>(10, 2);
+      weightNorm = new WeightNorm<>(linear);
+
+      model->Add(weightNorm);
+      model->Add<LogSoftMax<> >();
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      double error = model->Evaluate(model->Parameters(), 0, 1);
+      model->Gradient(model->Parameters(), 0, gradient, 1);
+      return error;
+    }
+
+    arma::mat& Parameters() { return model->Parameters(); }
+
+    FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>* model;
+    WeightNorm<>* weightNorm;
+    arma::mat input, target;
+  } function;
+
+  /*
+   * The following is the code similar to CheckGradient() function in
+   * ann_test_tools.hpp.
+   * For weightNorm the weights of wrapped layer are always calculated w.r.t
+   * the scalar and vector parameter before forwarding. Hence, it's gradient
+   * checking only depends on gradients of scalar and vector parameters. Thus,
+   * an offset is required for checking.
```

Comment:
If I understand correctly, then you do not need the gradient for the weights of `wrappedLayer` but only the gradient for the parameters for the weight norm layer.  Could you store the weights of the wrapped layer separately in `wrappedLayerWeights` and avoid calling `Gradient` for the wrappedLayer and computing gradient for its weights.

Reply:
No that's quite incorrect.
I need to calculate gradients of the weight's of the `wrappedLayer` in order to calculate the `gradients` of the parameter of the weight norm layer. Just look into `section 2.1`  in the below link.


> Here's the link for the paper.
> https://papers.nips.cc/paper/6114-weight-normalization-a-simple-reparameterization-to-accelerate-training-of-deep-neural-networks.pdf

Reply:
Yes, I see. But could you store the weights separately in `wrappedLayerWeights` because the `optimizer` would not need to update the weights of the wrapped layer and it should not because there is no use.

Reply:
Yes, you are quite correct about updating the weights. 
But if we need to keep `bias term` separate then we will need to update it using the `optimizer`. 
Hence, I am currently thinking to make the required changes once I finalize whether to keep the `bias` term unaffected.
Thank for the review. :+1: